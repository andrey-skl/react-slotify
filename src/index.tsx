import * as React from 'react';

interface RendererProps {
  childs: React.ReactNode;
}

type RendererType<P> = React.FC<RendererProps & P>;

interface SlotProps<P> {
  showChildren?: boolean;
  restProps?: P
}

type NormalOrFunctionChildren<P> = React.ReactNode | ((props: P) => React.ReactNode);

type SlotType<P> = {
  (props: SlotProps<P> & {children?: NormalOrFunctionChildren<P>}): React.ReactElement<any, any> | null;
  Renderer: RendererType<P>;
};

export function createSlot<P extends Object>(): SlotType<P> {

  const Slot: SlotType<P> = ({ children, showChildren, restProps }) => {
    if (!showChildren) {
      return null;
    }
    if (typeof children === 'function' && restProps) {
      return children(restProps);
    }
    return <>{children}</>;
  };

  const Renderer: RendererType<P> = ({ childs, children, ...restProps }) => {
    const slotted = React.Children.toArray(childs).find(child => {
      return React.isValidElement(child) && child.type === Slot;
    });

    if (!slotted || !React.isValidElement(slotted)) {
      return <>{children}</>;
    }
    return React.cloneElement(slotted, {
      showChildren: true,
      restProps,
    } as unknown as SlotProps<P>);
  };

  Slot.Renderer = Renderer;
  return Slot as SlotType<P>;
}
