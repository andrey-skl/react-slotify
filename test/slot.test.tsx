import * as React from 'react';
import renderer from 'react-test-renderer';

import { createSlot } from '../src';

describe('Slot', () => {
  it('Should render to slot', () => {
    const TestSlot = createSlot();

    const Component: React.FC = ({ children }) => {
      return (
        <div>
          beforeslot <TestSlot.Renderer childs={children} /> afterslot;
          ownChildren: {children}
        </div>
      );
    };

    expect(
      renderer
        .create(
          <div>
            beforecomponent
            <Component>
              BEFORE_SLOT<TestSlot>IN_SLOT</TestSlot>AFTER_SLOT
            </Component>
            aftercomponent
          </div>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it('Should render default content if slot is not used', () => {
    const TestSlot = createSlot();

    const Component: React.FC = ({ children }) => {
      return (
        <pre>
          beforeslot
          <TestSlot.Renderer childs={children}>DEFAULT_SLOT_CONTENT</TestSlot.Renderer>
          afterslot; ownChildren: {children}
        </pre>
      );
    };

    expect(
      renderer
        .create(
          <main>
            beforecomponent
            <Component>Slot not used</Component>
            aftercomponent
          </main>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
