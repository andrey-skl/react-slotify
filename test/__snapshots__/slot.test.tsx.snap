// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot Should pass default children to function 1`] = `
<main>
  <pre>
    <div>
      foo=
      DEFAULT_SLOT_CONTENT
    </div>
  </pre>
</main>
`;

exports[`Slot Should pass parameters 1`] = `
<main>
  <pre>
    <div>
      foo=
      bar
    </div>
  </pre>
</main>
`;

exports[`Slot Should render default content if slot is not used 1`] = `
<main>
  beforecomponent
  <pre>
    beforeslot
    DEFAULT_SLOT_CONTENT
    afterslot; ownChildren: 
    Slot not used
  </pre>
  aftercomponent
</main>
`;

exports[`Slot Should render to slot 1`] = `
<div>
  beforecomponent
  <div>
    beforeslot 
    IN_SLOT
     afterslot; ownChildren: 
    BEFORE_SLOT
    AFTER_SLOT
  </div>
  aftercomponent
</div>
`;
